data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,610 @@
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+The first version takes two extra parameters: the two previously computed Fibonacci numbers.
+
+The second version uses the two first Fibonacci numbers as initial values. There is nothing to be computed when returning from a recursive call, so this enables the compiler to perform an optimisation called “tail call elimination.” This turns recursivity into imperative iteration in the generated native code and leads to much-improved performances.
+
+## Multiple Arguments Functions
+
+### Arrow Types and Mutiple Arguments
+
+Until now, functions with several arguments had a type looking like this: `a₁ -> a₂ -> a₃ -> ⋯ -> b`
+
+Where `a₁` is the type of the first argument, `a₂` is the type of the second argument, etc., and `b` is the type of the result. However, this type isn't displayed as it is. In reality, this type is the following: `a₁ -> (a₂ -> (a₃ -> ⋯ -> b))`. The arrow symbol is a [binary operator](https://en.wikipedia.org/wiki/Binary_operation). Here is how this can be observed:
+```ocaml
+# ( + );;
+- : int -> int -> int = <fun>
+
+# let f : (int -> (int -> int)) = ( + );;
+val f : int -> int -> int = <fun>
+```
+
+In this example, the type of integer addition function `( + )` is displayed first. It prints as `int -> int -> int`, meaning it takes two integers and returns an integer. In the second statement, `f` is defined to have type `int -> (int -> int)`. Despite the type of `( + )` looking different, the binding with `f` is allowed because those two types are the same. The toplevel always prints the version without parentheses to make it easier to read.
+
+Putting the parentheses the other way does not work:
+```ocaml
+# let g : ((int -> int) -> int) = ( + );;
+Error: This expression has type int -> int -> int
+       but an expression was expected of type (int -> int) -> int
+       Type int is not compatible with type int -> int
+```
+
+In mathematical language, the type arrow operator _associates to the right_. Function types without parentheses should be treated as if they have parentheses placed to the right in the same way that the type of `f` was declared above. These types are the same:
+- `int -> int -> int`
+- `int -> (int -> int)`
+
+But only the first is displayed, so it is not the same as `(int -> int) -> int`.
+
+Most importantly, this means a *binary* function isn't really one that takes two parameters. In the case of addition, it is a function that takes an integer and returns a function of type `int -> int`. And it is indeed!
+```ocaml
+# ( + ) 2;;
+- : int -> int = <fun>
+```
+
+Passing a single integer to the addition returns a function of type `int -> int`. That function value is a closure. The value passed as the first parameter, here 2, is captured as if it had been in an earlier definition.
+
+### Passing Tuples
+
+Consider this function:
+```ocaml
+# let space_cat (s, t) = s ^ " " ^ t;;
```

Comment:
Why are there parentheses around s & t? Before function parameters have been written without ( ) ... because it's a string? ... or because of what's said in 499?